crypto/evp/pkey_mac.c

```diff
@@ -81,7 +82,8 @@ static int pkey_mac_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src)
     sctx = EVP_PKEY_CTX_get_data(src);
     dctx = EVP_PKEY_CTX_get_data(dst);
 
-    if (!EVP_MAC_CTX_copy(dctx->ctx, sctx->ctx))
+    if (sctx->ctx->data == NULL
+        || !dctx->ctx->meth->copy(dctx->ctx->data, sctx->ctx->data))
```

Comment:
This seems to be the only spot, other than in `mac_lib.c`, where you use the copy method function directly.  Would you consider some form of this instead?

``` C
    EVP_MAC_CTX_free(dctx->ctx);
    if ((dctx->ctx = EVP_MAC_CTX_dup(sctx->ctx)) == NULL)
        goto err;
```

That would allow you to convert the EVP_MAC copy function to a dup function.

Reply:
I find the sequence a little strange in that case:
```
ctx = malloc();
free(ctx);
ctx = dup();
```

Which is why I didn't do that.

Reply:
Yup, I know.  That's the exact reason why I *didn't* implement dup semantics, because it caused this type of strangeness, since the rest of OpenSSL uses copy semantics.  That's the cost of bridging between the two ways.

Of course, we can keep it this way, but then I find it strange to see dup semantics at the EVP_MAC API level, but still have a copy function in the EVP_MAC method.  As a method author, I'd raise an eye-brow and wonder what the hell is going on.  That's also a price to pay, although different.

Reply:
So I guess that leaves you with a choice, and it would be good if it was documented somehow.

Reply:
The external API we are implementing here is *dup*, this function is called by EVP_PKEY_CTX_dup(). That is, we allocate a new one with init(), and the copy() it.

Reply:
Well, we don't allocate with init(), but we call it to initialize things. Which for a dup function doesn't actually make sense, since you want to overwrite whatever init() did.

Reply:
Oooohhhhh...  hadn't payed attention to `EVP_PKEY_CTX_dup()`, and this gets quite confusing since all *other* EVP context types use copy in their public API :-/